### PR TITLE
Byte-encode Rmd source to retain utf-8 encoding

### PR DESCRIPTION
**Pull request by aoles**
Fixes the issue that utf-8 encoded special characters in the downloadable Rmd source file get scrambled, see https://github.com/rstudio/rmarkdown/issues/722.

### CODE REVIEW

**Comment by jjallaire** [comment]
LGTM. @jmcphers and @yihui any concerns here?

**Comment by yihui** [comment]
I have no idea. I'm not an expert here.

**Comment by jjallaire** [comment]
Okay, @jmcphers and @jcheng5 could you take a look and let us know if you see any problem with this (I confess it also out of my scope to know whether there are subtle problems which could be introduced here).

**Comment by jcheng5** [comment]
It's weird that this works. Technically `charCodeAt` can return up to 65535, while we're only assigning into an uint8 array. However `window.atob` returns not a proper UTF-8 string (where `ś` would be a single character) but instead all the bytes are treated as individual characters.

In the example you gave, `ś` is represented like this in the `src` variable:

```
String.fromCharCode(0xC5) + String.fromCharCode(0x9B)
```

Instead the "correct" representation would be:

```
String.fromCharCode(347)
```

In other words, this code works but only because two "wrongs" cancel each other out. If we test this with some CJK characters and it works, I'm fine with shipping it, but let's keep on the lookout for further strangeness. We can use something like [this](https://github.com/beatgammit/base64-js/blob/master/lib/b64.js) to avoid the string at all, and just go straight from base64 to TypedArray, if necessary.

**Comment by jcheng5** [comment]
Another way to go would be to forego JavaScript entirely and do something like:

```html
<a download="whatever.Rmd" href="data:text/x-r-markdown;charset=utf-8;base64,LS0tCnRpdGxlOiAiVGVzdGZpbGUiCmF1dGhvcjogIkFuZHJ6ZWogT2xlxZsiCm91dHB1dDoKICBodG1sX2RvY3VtZW50OgogICAgY29kZV9kb3dubG9hZDogeWVzCiAgICBzZWxmX2NvbnRhaW5lZDogbm8KLS0tCiAgCiAgVGhlIGRvY3VtZW50IGF1dGhvciBuYW1lIGNvbnRhaW5zIGFuIGFjY2VudC4K">...</a>
```

The download attribute is [supported in Chrome, Firefox, and Edge](http://caniuse.com/#feat=download). It isn't supported in Safari, but the results are no different than FileSaver.js (and note that I added the charset to the data URI, which improves matters for Safari).

**Comment by jjallaire** [comment]
Thanks @jcheng5, I'll look at those alternatives.

@aoles <https://github.com/aoles> note that it may be a bit before I look
at this further (too many things to do before useR!)

**Comment by aoles** [comment]
Awesome, thank you @jjallaire and @jcheng5 for looking into this!

**Comment by aoles** [comment]
Hi, I just wanted to quickly check with you guys the status of this issue. It has already been a while since I filed https://github.com/rstudio/rmarkdown/issues/722, but it doesn't seem to be resolved yet.

Cheers,
Andrzej
